[SYCL][UR] Update unified-runtime

jandres742 opened the pull request:
Update to

https://github.com/oneapi-src/unified-runtime/commit/61cb864b3a73918124bff11f9ad58dafb8c7769a

Signed-off-by: Jaime Arteaga <jaime.a.arteaga.molina@intel.com>

jandres742 commented:
SYCL / Linux / Build + LIT (pull_request_target)  https://github.com/intel/llvm/actions/runs/4154130558/jobs/7187700934 failing with:

```cpp
_deps/unified-runtime-src/source/loader/ur_loader.cpp:36:2: error: extra ';' [-Werror=pedantic]
   36 | };
      |  ^
_deps/unified-runtime-src/source/loader/ur_loader.cpp:43:2: error: extra ';' [-Werror=pedantic]
   43 | };
```

Posted PR in UR https://github.com/oneapi-src/unified-runtime/pull/254

igchor commented on sycl/plugins/unified_runtime/ur/ur.hpp:
```diff
@@ -356,13 +355,13 @@ class UrReturnHelper {
 // Returns plugin specific error and warning messages
 // TODO: promote to Unified Runtime API
-zer_result_t zerPluginGetLastError(char **message);
+ur_result_t zerPluginGetLastError(char **message);
```
I think you don't need this function anymore. You can just implement `urGetLastResult`.
Feel free to ignore this comment if this is something you plan to do in next PRs.

jandres742 replied:
Thanks @igchor . good point. will remove->implement in follow up patch.

jandres742 commented:
I have identified some issues in tests. Fixing them in my machine and will repost.

jandres742 commented:
@smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415

All I see is are errors like:

rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

smaslov-intel commented:
> @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> 
> All I see is are errors like:
> 
> rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

I don't see either. Did you try to restart it?
@bader : any idea?

bader commented:
> > @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> > All I see is are errors like:
> > rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied
> 
> I don't see either. Did you try to restart it? @bader : any idea?

No.

bader commented:
> @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> 
> All I see is are errors like:
> 
> rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

@intel/dpcpp-devops-reviewers, could you investigate, please? To reduce the impact on CI jobs we probably should disable broken systems in GitHub Actions.

smaslov-intel commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -1,31 +1,63 @@
+
+message(STATUS "UNIFIED_RUNTIME_LIB_NAME ${UNIFIED_RUNTIME_LIB_NAME}")
+
 add_library (UnifiedRuntime-Headers INTERFACE)
+add_library (UnifiedRuntimeLoader-Headers INTERFACE)
```
do we really need two different headers interfaces here? why?

jandres742 replied:
please see above.

smaslov-intel commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -1,31 +1,63 @@
 endif()
 
+add_library (UnifiedRuntimeLoader INTERFACE)
+get_filename_component(UNIFIED_RUNTIME_LIBRARY_SRC "${UNIFIED_RUNTIME_LOADER_SOURCE_DIR}" DIRECTORY)
```
`UNIFIED_RUNTIME_LIBRARY_SRC` seems unused

jandres742 replied:
@smaslov-intel : I added several changes here to start using the loader. ur_loader.so is now being created but not used yet. These extra instructions are me trying to use the ur_loader, but decided to focus on code and then continue with the loader later.

I can remove all these if you'd like.

smaslov-intel commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -52,9 +84,30 @@ add_sycl_plugin(unified_runtime
     LevelZeroLoader
 )
+
+if (TARGET loader)
```
I assume this `loader` target is from UR repo, right? Is it possible to name it so it is more clear?

jandres742 replied:
@smaslov-intel loader is the name being used at this moment in the UR. I could submit a PR to the UR to change the name.

jandres742 replied:
@smaslov-intel : used alias now, please see.

igchor commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -1,31 +1,63 @@
     "${UNIFIED_RUNTIME_INCLUDE_DIR}"
 )
 
+target_include_directories(UnifiedRuntimeLoader-Headers
```
I'm not sure if those `target_include_directories` and `target_link_libraries` and those extra variables (UNIFIED_RUNTIME_LOADER_SOURCE_DIR, UNIFIED_RUNTIME_INCLUDE_DIR, etc.) are actually needed.

UR provides a target named unified-runtime::loader which already defines include dirs, so I think it should be enough to just link with `unified-runtime::loader` (put it under `LIBRARIES`  in `add_sycl_plugin` below)

jandres742 replied:
same as above.

igchor commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -52,9 +84,30 @@ add_sycl_plugin(unified_runtime
     UnifiedRuntime-Headers
+    UnifiedRuntimeLoader-Headers
+    UnifiedRuntimeLoader
     LevelZeroLoader-Headers
```
I think we should link only with the UnifiedRuntimeLoader here. And the L0 adapter should be build as a standalone SHARED library that links with LevelZero loader. 

Something like this:
```
add_sycl_plugin(unified_runtime
  SOURCES
    "pi_unified_runtime.hpp"
    "pi_unified_runtime.cpp"
    LIBRARIES
    Threads::Threads
    UnifiedRuntimeLoader
    UnifiedRuntime-Headers
)
```

and:
```
add_library(ur_level_zero
    SHARED
        ${CMAKE_CURRENT_SOURCE_DIR}/ur_level_zero.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/ur_level_zero.hpp
        "../../../ur_bindings.hpp"
        "../../../pi2ur.hpp"
        "../../../pi2ur.cpp"
        "../../ur.hpp"
        "../../ur.cpp"
)

target_include_directories(ur_level_zero
  PRIVATE "${LEVEL_ZERO_INCLUDE_DIR}"
)

target_link_libraries(ur_level_zero
  PRIVATE
    LevelZeroLoader-Headers
    LevelZeroLoader
)
```

jandres742 replied:
ur_loader work to continue in https://github.com/intel/llvm/pull/8423

jandres742 commented:
the changes in the CMakeLlists for the loader are not  needed for the update and rest of changes. I could separate this into two patches, one for the changes in CMakeLists for the loader, and other for the changes in code.

@smaslov-intel , @igchor : what do you think?

jandres742 commented:
> the changes in the CMakeLlists for the loader are not needed for the update and rest of changes. I could separate this into two patches, one for the changes in CMakeLists for the loader, and other for the changes in code.
> 
> @smaslov-intel , @igchor : what do you think?

@smaslov-intel , @igchor : I have separated the changes in CMake for the loader in https://github.com/intel/llvm/pull/8423. I will incorporate your comments there, and this PR will be mostly about updating UR version.

smaslov-intel reviewed (approved):
Thanks! LGTM

> [jandres742](https://github.com/jandres742) [force-pushed](https://github.com/intel/llvm/compare/d580073eb4457f93055116a2e85a47da92cf09e4..45bd81e5248a5619030837a85253720221dda4d0) the updateur branch 2 times, most recently from [d580073 ](https://github.com/intel/llvm/commit/d580073eb4457f93055116a2e85a47da92cf09e4)to [45bd81e ](https://github.com/intel/llvm/commit/45bd81e5248a5619030837a85253720221dda4d0)[14 hours ago](https://github.com/intel/llvm/pull/8315#event-8579104475)

@jandres742: please try to avoid **force-push** in future as it makes incremental review impossible

jandres742 commented:
A first stable tag has been added to the UR. 

I need in any case to rebase due to conflicts. Should I:

1. rebase as is, and then submit a new patch updating to the stable tag, or?
2. update to stable tag and use this same PR?

I'm thinking to update to that commit and use that for now while we work on updating L0 adapter.

Please let me know what everyone thinks.

@smaslov-intel , @bader  @igchor

bader commented:
> I need in any case to rebase due to conflicts.

Please, use `git merge <remote>/sycl` command to resolve merge conflicts. 

To update the tag/source/<make any other change> just push another commit to your branch. 
Don't worry about the history in your branch. All commits in this PR will be squashed to keep the linear history in the `sycl` branch. 
https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request

jandres742 commented:
Rebased and updated to UR tag https://github.com/oneapi-src/unified-runtime/tree/weekly-2023-02-24

Idea is to use this stable tag for development for now.

steffenlarsen reviewed (approved):
Just a small nitpick, otherwise changes look good to me.

jandres742 commented on sycl/plugins/unified_runtime/CMakeLists.txt:
```diff
@@ -58,3 +71,7 @@ add_sycl_plugin(unified_runtime
+
+if (TARGET ${PROJECT_NAME}::ur_loader)
+  set_target_properties(hello_world PROPERTIES EXCLUDE_FROM_ALL 1 EXCLUDE_FROM_DEFAULT_BUILD 1)
+endif()
```
thanks @steffenlarsen . changing.

igchor reviewed (approved):
LGTM
